Extension review 2

hmtinc opened the pull request:
#### Popper Basic Usage 
``` js
//Create Basic Popper
var popperTest = cy.nodes()[0].popper({
  target: "targetID",
  popper: popperOptions,
  boundingBox : boundingBoxFunction
});
```
* Returns a Popper Object  
* (Required) Target : Refers to the actual HTML content of your popper
* (Optional) Options : Refer to [Popper.js](http://popper.js.org) for more information on popper options
* (Optional) BoundingBox : Provides ability to manually set a bounding box

#### Popper Ref Basic Usage
``` js 
cy.elements().nodes()[0].popperRef(options)
```

* Returns a Reference Object

#### Changes Made 

* Extension now no longer uses the scratch pad, so  multiple poppers per element can be created 
* popper() now returns a popper object allowing for the user to manage popper objects
* popperRef() API function added to allow for the generation of reference objects
* Collections Implementation now enforces the requirement that the extension only be used on singular elements by just applying operations on the first element `eles.popper => eles[0].popper`
* Popper.js is now included in the package.json as a dependency 
* Options Function that generated ID has been removed as it was unused
* Users can now provide their own bounding box to override the internally generated one `ele.popper({ .. , boundingBox : boundingBox})`
* Users can now provide their own Reference Object to override the internally generated one `ele.popper({ .. , refObject : refObject})`
* Extension defaults to {w:3, h:3}, if no overrides are provided
* File names have been renamed to make the structure of the extension more clear i.e `helper.js => createBoundingBox.js` 
* Reference Object function has been split from them index file to also make structure more clear
* Demo CSS is now inline in the Demo HTML to avoid any possible confusion 
* Demo HTML commented to describe key portions of the code 
* Demo HTML now has demos to sample the new override features

d2fong commented on src/core/render.js:
```diff
@@ -0,0 +1,41 @@
+//Include helper functions and Popper
+import Popper from 'popper.js';
```
Should this use require or does it need to be imported?
```js
const Popper = require('popper.js')
```

hmtinc replied:
Not sure if there is a way to get to work with require. But it seems the slight differences between require  and import cause an error to be thrown. 

This throws a type error 
```
var Popper = require('popper.js');
var p = new Popper(refObject, target, options) 
```

This succeeds. 
```
import Popper from 'popper.js';
var p = new Popper(refObject, target, options);
```

maxkfranz replied:
We should be using `require()` as a rule until `import` is well supported in Node.  Node 8 / npm 5 is still a bit iffy (behind a flag and only using the `.mjs` extension).  Maybe in a year or two, `import` will be practical for new projects.

`const Popper = require('popper.js')` seems to work for me on node@6.  I would expect webpack to be the same.  You could always `console.log(Popper)` to see what you've got.

maxkfranz commented on src/core/render.js:
```diff
@@ -0,0 +1,41 @@
+const createReferenceObject = require('./createReferenceObject.js');
+
+//Create a new popper object associated with a cytoscape element (Nodes or Edges)
+module.exports.createPopperObject = function (cyElement, userOptions) {
```
It may make sense to pull out all the if-node, if-edge, if-cy bits and just have the common stuff here.  The cy stuff would go in `coreRef()` and ele stuff goes in `collectionRef()`.  You can use `ele.isNode()` in `collectionRef()` if you need

maxkfranz commented on src/core/index.js:
```diff
@@ -1,7 +1,75 @@
+
+  //Popper.js Should only be used on 1 element
+  if (elements.length > 1) {
+    console.log("Popper.js Extension should only be used on one element.");
```
console.warn() may be a better fit

https://developer.mozilla.org/en-US/docs/Web/API/Console/warn

maxkfranz commented on src/core/createReferenceObject.js:
```diff
@@ -0,0 +1,36 @@
+
+//Create a reference object for an element
+module.exports.getRef= function (cyElement, userOptions) {
+    //Determine element properties to determine how to generate a reference object
```
Similarly to a later comment, functions like these should be generic.  They could accept functions as arguments if you need customisation per type.

hmtinc commented:
@maxkfranz  

I have the made the following changes 

* Included an option to pass in a HTML element, so you don't have to specify an ID. 
* Added an example in the demo file to display the ability to pass in a HTML Element as the target
* Changed `throw msg;` to `throw new Exception(msg);`
* Removed all redundant variable checks (isNode, isCy, isElement, etc) from render.js
* Updated warning messages to use `console.warn` instead of `console.log`
* Removed build file/.DS_Store, and ignored them in gitignore

maxkfranz commented on src/core/createReferenceObject.js:
```diff
@@ -0,0 +1,36 @@
+                return createBoundingBox.getPopperBoundingBox(cyElement, cy, isNode, dim, userOptions.boundingBox);
+            },
+            get clientWidth() {
+                return dim.w;
```
May need to generate dim every call to `clientWidth()` in order to make it dynamic.  Consider a node that has a size change while the popper is on it

hmtinc commented:
@maxkfranz  I changed the referenceObject functions to dynamically create a dim on each call.

maxkfranz commented:
OK, this looks like a good base.  I'll merge this in and I'll create tickets for potential improvements
